### PR TITLE
std: refactor test-runner to no longer require tests to be exported

### PR DESCRIPTION
**Pull request by Yethal**
# Description

Test runner now performs following actions in order to run tests:
* Module file is opened
* Public function with random name is added to the source code, this function calls user-specified private function
* Modified module file is saved under random name in $nu.temp-path
* Modified module file is imported in subprocess, injected function is called by the test runner
# User-Facing Changes

* Test functions no longer need to be exported
* test functions no longer need to reside in separate test_ files
* setup and teardown renamed to before-each and after-each respectively
* before-all and after-all functions added that run before all tests in given module. This matches the behavior of test runners used by other languages such as JUnit/TestNG or Mocha
# Tests + Formatting

# After Submitting

### CODE REVIEW

File: crates/nu-std/tests/test_dirs.nu
Comment (skelly37) [inline crates/nu-std/tests/test_dirs.nu:7]: I think the replacement tool went wrong

File: crates/nu-std/std/testing.nu
Comment (skelly37) [inline crates/nu-std/std/testing.nu:241]:
```suggestion
            $"(ansi purple)some tests did not pass (char lparen)see complete errors below(char rparen):(ansi reset)"
```

File: crates/nu-std/std/testing.nu
Comment (skelly37) [inline crates/nu-std/std/testing.nu:199]: You can't just iterate through `ls` output here. You have to check the whole directory tree. In the current state, logger tests are not executed, as they are in a separate subdirectory

Comment (Yethal) [reply crates/nu-std/std/testing.nu:199]: I didn't write this, this is part of the original design

**Comment by amtoine** [comment]
@Yethal 
i see you're still pushing to this branch so, as we're releasing tomorrow, i'll
- make this a draft to let you finish
- mark it as `wait-for-after-nushell-release` to have a look later

i'll come to this for sure after the release :wink:

**Comment by Yethal** [comment]
@amtoine This was the last push actually, it's done now.

**Comment by amtoine** [comment]
> @amtoine This was the last push actually, it's done now.
>
ooh ok, great, so we can forget the first part of my comment :wink:

File: crates/nu-std/tests/test_dirs.nu
Comment (skelly37) [reply crates/nu-std/tests/test_dirs.nu]: I mean actually it happened in the whole file, not just here

File: crates/nu-std/std/testing.nu
Comment (skelly37) [reply crates/nu-std/std/testing.nu:199]: Sure, works just fine right now, might be closed

**Review by amtoine — COMMENTED** [review]
i'm not sure i'll have that much time tonight, so i've looked at all the files but `testing.nu` and i've got a little question :yum: 

- :question: out of curiosity, why were `setup` and `teardown` renamed to `before-each` and `after-each`?

File: crates/nu-std/std/testing.nu
Comment (amtoine) [inline crates/nu-std/std/testing.nu:222]: here we might directly compute the `tests` column without the intermediate `commands` one :yum: 
```bash
    let modules = (
        ls ($path | path join $module_search_pattern)
        | each {|row| {
            file: $row.name
            name: ($row.name | path parse | get stem)
        }}
        | insert tests {|module|
            get-commands $module.file | where $it starts-with "test_"
        }
        | filter {|x| ($x.tests | length) > 0}
        | filter {|x| if ($test | is-empty) { true } else { $test in $x.tests }}
        | filter {|x| if ($module | is-empty) { true } else { $module == $x.name }}
        | upsert before-each {|module| "before-each" in $module.tests}
        | upsert before-all {|module| "before-all" in $module.tests}
        | upsert after-each {|module| "after-each" in $module.tests}
        | upsert after-all {|module| "after-all" in $module.tests}
        | rename file name tests
        | update tests {|x|
            if ($test | is-empty) {
                $x.tests
            } else {
                $x.tests | where $it == $test
            }
        }
    )
```
> **Note**
> there is a tiny bit of formatting too :relieved:

Comment (Yethal) [reply crates/nu-std/std/testing.nu:225]: This won't pick up before-each and after-each commands as they're not prefixed with test_ and that's what you're inserting

**Comment by Yethal** [comment]
@amtoine setup and teardown were renamed to before-each and after-each to make them consistent with before-all and after-all as those are the names used in test runners in other languages such as Mocha or JUnit/TestNG. I can rename them back if necessary but then we'll have to come up with better names for before-all and after-all.

File: crates/nu-std/std/testing.nu
Comment (Yethal) [reply crates/nu-std/std/testing.nu]: fixed

File: .github/workflows/ci.yml
Comment (Yethal) [reply .github/workflows/ci.yml]: fixed

**Comment by skelly37** [comment]
nUnit uses `SetUpFixture` if we really want to stick to the older name. Yet I agree with @Yethal that before-all/each looks better

File: crates/nu-std/std/testing.nu
Comment (amtoine) [reply crates/nu-std/std/testing.nu:221]: this is a very good and valid point, thanks for the clarification about the `before_` and `after_` commands :relieved:

**Review by amtoine — APPROVED** [review]
@Yethal @skelly37 
thanks for the hints about the `before_` and `after_` commands, that sounds fine with me :ok_hand: 

@Yethal 
thanks for the work in this PR, this looks great :partying_face:

**Comment by amtoine** [comment]
## :postal_horn: :postal_horn: little annoucement
the *Typos* CI job is failing in this PR.
this has been fixed in https://github.com/nushell/nushell/pull/9393 :+1: 

in this PR, you can either
- merge the new `main` branch into the PR branch to explicitely have a :heavy_check_mark: here
- wait for the PR to land and the CI should be :heavy_check_mark: once landed on the `main` branch :relieved:

**Comment by amtoine** [comment]
let's make sure the CI is green and land this... :recycle:

**Comment by amtoine** [comment]
can't wait to see your `#[test]` annotation PR @Yethal :star_struck:

**Pull request merged**
